Src/Support/Google.Apis.Auth/OAuth2/Flows/AuthorizationCodeFlow.cs

```diff
@@ -310,11 +310,17 @@ public virtual Task RevokeTokenAsync(string userId, string token, CancellationTo
             catch (TokenResponseException ex)
             {
                 // In case there is an exception during getting the token, we delete any user's token information from 
-                // the data store.
-                tokenException = ex;
+                // the data store if it's not a server-side error.
+                int statusCode = (int)(ex.StatusCode ?? (HttpStatusCode)0);
+                bool serverError = statusCode >= 500 && statusCode < 600;
+                if (!serverError)
+                {
+                    // If not a server error, then delete the user token information.
+                    // This is to guard against suspicious client-side behaviour.
+                    await DeleteTokenAsync(userId, taskCancellationToken).ConfigureAwait(false);
+                }
+                throw;
             }
-            await DeleteTokenAsync(userId, taskCancellationToken).ConfigureAwait(false);
-            throw tokenException;
```

Comment:
Ah, this is presumably outside the catch block because of the pre-C#6 restriction on awaiting in catch blocks. Urgh.